### ISSUE TITLE
A.09.IR04.srv.linkage

### ISSUE THREAD

**Issue by jensscheerlinck** [inspire-eu-validation/metadata#55]
*This issue has been extracted from the issue list on:https://ies-svn.jrc.ec.europa.eu/issues/2685*
# Comment
Ambiguity: 'The URL is resolved.' It is not entirely clear what resolving the URI means. Which maximum timeout can be applied? Should we only look at the HTTP status code (200), or also at the returned information?
Ambiguity: It is not entirely clear which parts of the WSDL or GetCapabilities document need to be tested: 'If the response indicates a linkage is a service capabilities or WSDL document, some basic params in the service response are analysed.' Or does it only refer to: 'Any service response should be checked if it provides proper linkage. The service wsdl or capabilities document should have a featuretype that shares the resource unique identification.'
Testability: A manual test is suggested, if the resource locator is a web page with further instructions or a client application. 
Note: Perhaps CI_OnLineFunctionCode (optional element) could be used to determine the nature of the online resource locator if present (see also TG SDS req 3).
Note: What in case of access restrictions?
Note: Another prerequisite is test case A.04.
# Proposed Change
Specify which HTTP status codes are acceptable.

**Comment by pvgenuchten** [comment]
These are good suggestions made for "resolving", since many tests reference "resolve", i added an explanation on https://github.com/inspire-eu-validation/ats-metadata/blob/master/README.md#resolve, would be good to add aspects such as statuscodes and timeouts.

A typical test to perform on getcapabilities is to test if one of the advertised layers/featuretypes has same identification as the metadata. 

In case of access restrictions, a manual test is required

**Comment by pvgenuchten** [comment]
Email discussion with @thijsbrentjens:

Thijs:  Some issues are on consistency between services and metadata documents. For example: check existence of dataset identifiers found in ATOM service feeds, in the referred Metadata documents (https://github.com/inspire-eu-validation/ats-download-atom/issues/63). Strictly, the ATS of a service is not on valid or consistent metadata documents so you can't be sure that a value is NOT present in the metadata (because the metadata might not be encoded correctly for example), but the tests can be useful to have. I remember we discussed an ATS for interoperability or something alike, in the face to face meeting in Ispra. But I can't recall what we decided then. Should we collect test cases that could be part of such an ATS? Do others have similar issues? Can we agree on a common approach?

Paul: Yes, this is an important question. The risk of each of the test suites only testing it’s own responsibilities is that all tests are successful, but still you don’t have a working system. In the metadata tests we have similar challenges, in cases that metadata is referring to services, you would want to verify the get capabilities document of that service, to check if it is indeed the described service (by checking if a layer has the same identification as the metadata). But who’s to blame if a service capabilities has a different identification then the describing metadata? Which of the two tests should fail (or both). Personally i’d say both, so it is a shared responsibility of the two.

Thijs:
I disagree with you on failing both ATSes: that would mean we have to perform the tests of another ATS too or better maybe: have a prerequisite to successfully pass the other ATS. For example: what means "valid metadata" if we are testing a network service's reference to a service metadata document? Pass the metadata ATS first? And how would that work if the ATS of metadata has a prerequisite to successfully pass the ATS of a network service? I think we need a separate ATS, for interoperability / consistency.

Paul: 
I agree with your statement that by introducing these cross dependencies into the tests, testing and evaluating the results will become much harder. However not testing the links at all is not an option imho. Some suggested to create a separate test suite to test linkage between services and metadata. Maybe this is what @michellutz wants to discuss (and implement) in https://github.com/inspire-eu-validation/ats-cross-cutting

**Comment by cportele** [comment]
First of all, right now we can only test consistency constraints, if they are specified in a TG. In some places this seems to be the case, but there is propably room to specify (and then test) additional consistency tests that are not covered today. Information in service capabilities and their metadata records is often very different.

Adding a "TG for interoperability" is an option, but I do not think it is necessary. In general it should be possible to determine the target for which the constraint should be tested (and in which TG the requirement should go). This will depend on logical aspects (what is the original / more authoritative value) and from test execution constraints (it does not make sense to download 100 GB of data to find out whether the character encoding stated in the metadata record is used - much easier to do it the other way round).

**Comment by aquaglia** [comment]
I agree that it is unreasonable to download the full dataset every time the metadata is validated.
It is also practically valueless since by the time all the dataset has been validated, it might have changed and become invalid.

However, it is feasible and in general useful to "probe" and test only the first bytes/kilobytes/megabytes of a dataset. For xml this is enough to determine if the character encoding, the spatial object type, the encoding, the transfer rate and so on, correspond to what is declared and/or required.

In addition, there are or there will be tests that make sense only once all the metadata documents of the INSPIRE Infrastructure (or for a Member State with more than one National Discovery Serice like Belgium) have been collected in a single, possibly virtual, container.

**Comment by pvgenuchten** [comment]
@cportele, can you explain a bit more: "In general it should be possible to determine the target for which the constraint should be tested". Do you mean that to solve this, the most relevant tg should be updated to include an improved check for linkage consistency, after which this can be implemented in of the ats's?

In this case either the service ats could verify the link to metadata, or the metadata ats could test the link to the service, to see if both use similar dataset identification(s)

**Comment by cportele** [comment]
@aquaglia "Probing" in general will not work. The only mandated download services are the pre-defined ones and I think you will have to download the complete archive, decompress it and then you can check, if all the files inside are using UTF-8.

@pvgenuchten Yes, that was my point, one test is sufficient, ie. the TG that is considered more appropriate should be updated whenever a missing consistenty constraint has been identified.

**Comment by aquaglia** [comment]
@cportele Curiously enough, I am actually using the probing technique in the Validator 2 and it seems to be working fine. INSPIRE predefined download services are being implemented either as Atom based services or as WFS with predefined stored queries.
In the WFS case, there is no problem because either you interrupt the download or you ask for less features.
In the case of Atom based services, I have noticed that either the download links are actually WFS calls, so we fall back to the previous case or the links point to zip files of small dimensions that can be fully downloaded an unzipped.
I believe that organisations prefer not to build huge zip files.
I think that, if we take the opportunity, while writing the ATS, to make probing possible, the validators are going to be much more useful and the INSPIRE infrastructure will be much more robust.

**Comment by cportele** [comment]
@aquaglia 
I am not convinced. With WFS there is the risk that even though you "interrupt" the download on the client side, the server will still complete the response (which will go nowhere). This may put a lot of strain on the server, if the response is implemented as a dynamic query and not as a pre-cached file.

Regarding Atom, you are saying that almost all datasets are making use of the multiple files option and all of the files are of a small size? Interesting. Are there any statistics available?

**Comment by aquaglia** [comment]
@cportele A request is not going to be any heavier if the client closes the connection before receiving the entire response, so I do not see a problem, here. It is quite a common thing for a large and long download to be interrupted, for many possible reasons.

I can try to put together some statistics.

**Comment by cportele** [comment]
Resolution depends on resolution of #27

**Comment by jensscheerlinck** [comment]
To summarize this issue: @pvgenuchten has added a suggestion for an explanation for "[resolve](https://github.com/inspire-eu-validation/ats-metadata/blob/master/README.md#resolve)" to the ATS readme.md.

Based on the discussion here between @pvgenuchten, @cportele and @aquaglia I did not see a clear consensus to improve un the current suggestion. 

Therefore I propose to mark this issue either as implemented or wontfix. Unless we can agree on introducing the concept of 'probing' or another solution to this issue.

**Comment by aquaglia** [comment]
I think there are two main points here:

1) The expression “The URL is resolved.” is used.
The word "resolved" is misleading because that, for a URL, usually means to resolve the domain name to an IP address.

2) What validation steps have to be performed in case of a service metadata.
If the service is a Network Service, i.e. a serviceType is specified and it is one among discovery, view, download, etc. then a resource locator must be present at which a known implementation of a Network Service responds. 
The ATS must also specify what is the outcome of the test if the Network Service does not answer, for whatever reason.
From the experience of the INSPIRE Geoportal I can say that it happens in extremely rare occasions.
It would be just reasonable for the test to fail and it will have to be repeated later.
How many times the test is allowed to fail in a specified time frame, that is a matter of Quality Of Service, so not to be dealt with in this Abstract Test.
Of course, in case of a Spatial Data Service which is not a Network Service, the situation gets much more complicated and needs to be described in this Abstract Test.
